### PR TITLE
Update documentation with new features and instructions for new blocks

### LINKED ISSUE #66: Document how to add new blocks
It would be nice to have a guide on how to add new blocks to the solver for new developers. I am working with @kharold23 and @aabrown100-git to add a new valve block. I will try and document the steps along the way and later add that to the documentation.

### PR DESCRIPTION
**Pull request by menon-karthik**
## Current situation
Closes #66 
We recently had a few new features implemented (or in soon-to-be-merged PRs). So I made some updates to the documentation to reflect those changes. 

## Release Notes 
* Added a new page documenting how to add new blocks.
* Added a new section on contributing to the code.
* Added documentation on #108 
* Added documentation on #109 
* Added short summary of available blocks (including new ones)

## Documentation
This is the documentation.

## Testing
Documentation builds. Did not change anything else.

## Code of Conduct & Contributing Guidelines 
- I agree to follow the [Code of Conduct](https://github.com/SimVascular/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/SimVascular/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: docs/pages/main.md [hunk 1]
```diff
@@ -314,13 +326,17 @@ Blood vessel \n junction              | MODEL::BloodVesselJunction  | `BloodVess
 
 Description                           | Class                       | `bc_type`             | `bc_values`
 ------------------------------------- | --------------------------- | --------------------- | ----------- 
-Prescribed (transient) flow           | MODEL::FlowReferenceBC      | `FLOW`                | `Q`: Time-dependent flow values \n `t`: Time stamps
-Prescribed (transient) pressure       | MODEL::PressureReferenceBC  | `PRESSURE`            | `P`: Time-dependent pressure values \n `t`: Time stamps
+Prescribed (transient) flow           | MODEL::FlowReferenceBC      | `FLOW`                | `Q`: Time-dependent flow values \n `t`: Time steps \n `fn`: Mathematical expression
```
Comment (mrp089) [inline docs/pages/main.md]: Indicate that it's either (`Q` and `t`) or `fn`

File: docs/pages/main.md [hunk 2]
```diff
@@ -314,13 +326,17 @@ Blood vessel \n junction              | MODEL::BloodVesselJunction  | `BloodVess
...
 Windkessel                            | MODEL::WindkesselBC         | `RCR`                 | `Rp`: Proximal resistance \n `C`: Capacitance \n `Rd`: Distal resistance \n `Pd`: Distal pressure
+Coronary outlet                       | MODEL::OpenLoopCoronaryBC   | `CORONARY`            | `Ra`: Proximal resistance \n `Ram`: Microvascular resistance \n `Rv`: Venous resistance \n `Ca`: Small artery capacitance \n `Cim`: Intramyocardial capacitance \n `Pim`: Intramyocardial pressure \n `Pv`: Venous pressure
+
+The above table describes the most commonly used boundary conditions. In addition, svZeroDSolver includes various closed-loop boundary conditions. Examples can be found in `svZeroDSolver/tests/cases`.
 
+Note that the `FLOW` and `PRESSURE` boundary conditions accept mathematical expressions in `bc_values`. For an example where values of the boundary condition are specified as a function of time, see `svZeroDSolver/tests/cases/pulsatileFlow_R_RCR.json`. For an example with a mathematical expression for the boundary condition, see `svZeroDSolver/tests/cases/timeDep_Flow.json`.
```
Comment (mrp089) [inline docs/pages/main.md]: Might be easier to read if the examples were included here (and easier to see for new users that mathematical expressions are possible).

File: docs/pages/developer_guide.md [hunk 3]
```diff
@@ -30,15 +26,35 @@ cmake -DCMAKE_BUILD_TYPE=Debug ..
...
 This is useful when continuously running the integration tests during development.
 
-## Code Style
+# Contributing to svZeroDSolver
+
+**NOTE: To contribute new developments to the main branch of svZeroDSolver, developers must first open an issue on the svZeroDSolver Github repository to describe the planned changes.** 
+
+* The changes should be implemented in a feature branch of the developer's fork of svZeroDSolver. 
+* Once the changes are implemented, the developer should make sure the build, documentation, and code format tests are passing on the user's feature branch. 
+  * The tests are automatically run when pushing changes to the developer's remote branch on Github.
```
Comment (mrp089) [inline docs/pages/developer_guide.md]: Good thing to write this down

File: docs/pages/add_block.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+<p> <br> </p>
+
+## 2. Create a class for the new block.
+
+### Class constructor
+
+* The new class will be inherited from `Block`. Define a constructor of the form:
+```
+    GenericBlock(int id, Model *model)
```
Comment (mrp089) [inline docs/pages/add_block.md]: Maybe something like `MyNewBlock` instead of `GenericBlock`

File: docs/pages/add_block.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    {{Param_1, InputParameter()}, 
+    {Param_2, InputParameter()}, 
+    ..., 
+    {Param_N, InputParameter()}}) {}
+```
+  * `GenericBlock` is the name of the new class
+  * `block_type` and `block_class` are the same as what was added in Step 1 above.
+  * The names of the input parameters of the block are `Param_1`, ... , `Param_N`. 
+  * The properties of each parameter are defined by `InputParameter`, which specifies whether it is optional, an array, a scalar, a function, and its default value.
+  * The names `Param_1`, ... , `Param_N` should be the same as the parameter names within the block definition in the `.json` configuration/input file.
```
Comment (mrp089) [inline docs/pages/add_block.md]: should or have to?

File: docs/pages/add_block.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+<p> <br> </p>
+
+### Set up the degrees of freedom
+* The class should have a `setup_dofs(DOFHandler &dofhandler)` function.
```
Comment (mrp089) [inline docs/pages/add_block.md]: should or has to?

File: docs/pages/add_block.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+<p> <br> </p>
+
+### An example
+
+* Assume a block has the following non-linear governing equations:
+```
+a*dQ_in/dt + b*P_in + c*(dP_in/dt)*Q_in + d = 0
+```
+```
+e*dP_out/dt + f*Q_out*Q_out + g*P_out + h*I_1 = 0
```
Comment (mrp089) [inline docs/pages/add_block.md]: Write in math mode?

File: docs/pages/add_block.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+<p> <br> </p>
+
+## 4. Implement the matrix equations for the block.
+
+* Implement the `update_constant`, `update_time` and `update_solution` functions.
+  * All matrix elements that are constant are specified in `update_constant`.
+  * Matrix elements that depend only on time (not the state variables) are specified in `update_time`.
+  * Matrix elements that change with the solution (i.e. depend on the state variables themselves) are specified in `update_solution`. 
+  * *Note: Not all blocks will require the `update_time` and `update_solution` functions.*
```
Comment (mrp089) [inline docs/pages/add_block.md]: Would change the language above from "should" to something like "must"

File: docs/pages/add_block.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+<p> <br> </p>
+
+* *Note: Any matrix and vector components that are not specified are 0 by default.*
+
+<p> <br> </p>
+
+## 4. Add the new block to the build system. 
+
+* Add `GenericBlock.h` and `GenericBlock.cpp` to `src/model/CMakeLists.txt`
```
Comment (mrp089) [inline docs/pages/add_block.md]: Would consistently move something like `MyNewBlock`

**Review by mrp089 — CHANGES_REQUESTED** [review]
Looks very nice! Thank you for updating the documentation! I added a few minor ideas.

**Comment by mrp089** [comment]
The `macos-11` GitHub runner is [not available anymore](https://github.com/actions/runner-images/issues/9255) as of July, so I removed it.

I added a `macos-latest` runner (and `ubuntu-latest`) so we always stay up to date. However, there seems to be a problem with `conda`. Up to you, @menon-karthik, if you want to do this upgrade now or leave it for a later PR.

**Comment by menon-karthik** [comment]
> The `macos-11` GitHub runner is [not available anymore](https://github.com/actions/runner-images/issues/9255) as of July, so I removed it.
> 
> I added a `macos-latest` runner (and `ubuntu-latest`) so we always stay up to date. However, there seems to be a problem with `conda`. Up to you, @menon-karthik, if you want to do this upgrade now or leave it for a later PR.

I think it's a good idea to stay ahead of the curve, so let's upgrade that right away. I fixed the `conda` problem by using [conda-incubator](https://github.com/conda-incubator/setup-miniconda) to find `conda` and add it to the path. Also, it looks like `macos-12` is also going to be removed soon. I will create a new issue and PR to upgrade that just so it is recorded.

**Review by mrp089 — APPROVED** [review]
